opm/simulators/linalg/cuistl/CuView.hpp

```diff
@@ -0,0 +1,395 @@
+     * @param data the vector to copy to
+     *
+     * @note This does synchronous transfer.
+     * @note This assumes that the size of this vector is equal to the size of the input vector.
+     */
+    void copyToHost(std::vector<T>& data) const;
+
+    /**
+     * @brief size returns the size (number of T elements) in the vector
+     * @return number of elements
+     */
+    __host__ __device__ size_t size() const{
+        // return std::size_t(m_numberOfElements);
+        return detail::to_size_t(m_numberOfElements);
+    }
+
+    /**
+     * @brief creates an std::vector of the same size and copies the GPU data to this std::vector
+     * @return an std::vector containing the elements copied from the GPU.
+     */
+    std::vector<T> asStdVector() const;
+    /// @brief Iterator class to make CuViews more similar to std containers
+    class iterator {
+    public:
+        // Iterator typedefs
+        using iterator_category = std::forward_iterator_tag;
+        using difference_type = std::ptrdiff_t;
+        using value_type = T;
+        using pointer = T*;
+        using reference = T&;
+
+        /// @brief Create iterator from a pointer
+        /// @param ptr provided pointer that will become an iterator
+        /// @return // the created iterator object
+        __host__ __device__ iterator(T* ptr) : m_ptr(ptr) {}
+
+        /// @brief Dereference operator
+        /// @return retrieve what the iterator points at
+        __host__ __device__ reference operator*() const {
+            return *m_ptr;
+        }
+
+        /// @brief Pre-increment operator
+        /// @return return the pointer after it is incremented
+        __host__ __device__ iterator& operator++() {
+            ++m_ptr;
+            return *this;
+        }
+
+        /// @brief Post-increment operator
+        /// @param no parameter, int is placeholder for c++ implementation to differentiate from pre-increment
+        /// @return Iterator before it is incremented
+        __host__ __device__ iterator operator++(int) {
+            iterator tmp = *this;
+            ++m_ptr;
+            return tmp;
+        }
+
+        /// @brief Pre-decrement operator
+        /// @return return the pointer after it is decremented
+        __host__ __device__ iterator& operator--() {
+            --m_ptr;
+            return *this;
+        }
+
+        /// @brief Post-decrement operator
+        /// @param no parameter, int is placeholder for c++ implementation to differentiate from pre-decrement
+        /// @return Iterator before it is decremented
+        __host__ __device__ iterator operator--(int) {
+            iterator tmp = *this;
+            --m_ptr;
+            return tmp;
+        }
+
+        /// @brief Inequality comparison operator
+        /// @return boolean value that is true if the pointers contains different addresses
+        __host__ __device__ bool operator!=(const iterator& other) const {
+            return !(*m_ptr == *other.m_ptr);
```

Comment:
Why are you dereferencing here? The pointers can be unequal even though the value they point to is the same.